src/common/attribute.rs

```diff
@@ -0,0 +1,626 @@
+		///
+		/// [`backing_store`]: Attributes::backing_store
+		const BACKING_STORE = 0x0000_0040;
+		/// See also: [`backing_planes`], [`u32`].
+		///
+		/// [`backing_planes`]: Attributes::backing_planes
+		const BACKING_PLANES = 0x0000_0080;
+		/// See also: [`backing_pixel`], [`Pixel`].
+		///
+		/// [`backing_pixel`]: Attributes::backing_pixel
+		const BACKING_PIXEL = 0x0000_0100;
+
+		/// See also: [`override_redirect`], [`bool`].
+		///
+		/// [`override_redirect`]: Attributes::OverrideRedirect
+		const OVERRIDE_REDIRECT = 0x0000_0200;
+		/// See also: [`save_under`], [`bool`].
+		///
+		/// [`save_under`]: Attributes::save_under
+		const SAVE_UNDER = 0x0000_0400;
+
+		/// See also: [`event_mask`], [`EventMask`].
+		///
+		/// [`event_mask`]: Attributes::event_mask
+		const EVENT_MASK = 0x0000_0800;
+		/// See also: [`do_not_propagate_mask`], [`DeviceEventMask`].
+		///
+		/// [`do_not_propagate_mask`]: Attributes::do_not_propagate_mask
+		const DO_NOT_PROPAGATE_MASK = 0x0000_1000;
+
+		/// See also: [`colormap`], <code>[CopyableFromParent]<[Colormap]></code>.
+		///
+		/// [`colormap`]: Attributes::colormap
+		const COLORMAP = 0x0000_2000;
+
+		/// See also: [`cursor_appearance`], [`CursorAppearance`].
+		///
+		/// [`cursor_appearance`]: Attributes::cursor_appearance
+		const CURSOR_APPEARANCE = 0x0000_4000;
+	}
+}
+
+/// A boolean type wrapping a `u32` value for use in [`Attributes`].
+///
+/// This is used to easily encode booleans as four bytes. It is not part of the
+/// public API.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, ConstantX11Size, X11Size, Readable, Writable)]
+struct Bool(u32);
+
+impl From<bool> for Bool {
+	fn from(value: bool) -> Self {
+		if value {
+			Self(1)
+		} else {
+			Self(0)
+		}
+	}
+}
+
+impl From<Bool> for bool {
+	fn from(value: Bool) -> Self {
+		value.0 != 0
+	}
+}
+
+/// A set of attributes for a [window].
+///
+/// [window]: Window
+#[derive(Debug, Hash, PartialEq, Eq)]
+pub struct Attributes {
+	/// Total [`X11Size`] of these `Attributes`.
+	///
+	/// This is cached so that it doesn't have to be recalculated each time -
+	/// `Attributes` is immutable.
+	///
+	/// This field is not part of the X11 format for this struct.
+	x11_size: usize,
+
+	mask: AttributeMask,
+
+	background_pixmap: Option<ParentRelatable<Option<Pixmap>>>,
+	background_pixel: Option<Pixel>,
+
+	border_pixmap: Option<CopyableFromParent<Pixmap>>,
+	border_pixel: Option<Pixel>,
+
+	bit_gravity: Option<BitGravity>,
+	win_gravity: Option<WinGravity>,
+
+	backing_store: Option<BackingStore>,
+	backing_planes: Option<u32>,
+	backing_pixel: Option<Pixel>,
+
+	override_redirect: Option<Bool>,
+	save_under: Option<Bool>,
+
+	event_mask: Option<EventMask>,
+	do_not_propagate_mask: Option<DeviceEventMask>,
+
+	colormap: Option<CopyableFromParent<Colormap>>,
+
+	#[allow(clippy::option_option)]
+	cursor_appearance: Option<Option<CursorAppearance>>,
+}
+
+impl Attributes {
+	#[must_use]
+	#[allow(clippy::too_many_arguments)]
```

Comment:
I don't understand this lint to be honest, maybe I'm missing something.  Yes, there are a lot of arguments... but there are a lot of fields... perhaps it means that you can't use struct update syntax or whatever since there are private fields.  I guess the only solution to that would be to have a `AttributesBuilder` struct with public fields that is used as the argument here... might be a good idea I suppose.